extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,321 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.S3Uri;
+import software.amazon.awssdk.services.s3.model.HeadObjectResponse;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.NonWritableChannelException;
+import java.nio.channels.SeekableByteChannel;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+
+/**
+ * {@link SeekableByteChannel} class used to fetch objects from AWS S3 buckets using an async client with the ability to
+ * read ahead and cache fragments of the object.
+ */
+final class S3SeekableByteChannel implements SeekableByteChannel, SeekableChannelsProvider.ContextHolder {
+
+    private static final long CLOSED_SENTINEL = -1;
+
+    /**
+     * Context object used to store read-ahead buffers for efficiently reading from S3.
+     */
+    static final class S3ChannelContext implements SeekableChannelsProvider.ChannelContext {
+
+        /**
+         * Used to store information related to a single fragment
+         */
+        static class FragmentContext {
+            /**
+             * The index of the fragment in the object
+             */
+            private int fragmentIndex;
+
+            /**
+             * The future that will be completed with the fragment's bytes
+             */
+            private CompletableFuture<ByteBuffer> future;
+
+            private FragmentContext(final int fragmentIndex, final CompletableFuture<ByteBuffer> future) {
+                this.fragmentIndex = fragmentIndex;
+                this.future = future;
+            }
+        }
+
+        /**
+         * Used to cache recently fetched fragments for faster lookup
+         */
+        private final List<FragmentContext> bufferCache;
+
+        private long size;
+
+        S3ChannelContext(final int maxCacheSize) {
+            bufferCache = new ArrayList<>(maxCacheSize);
+            for (int i = 0; i < maxCacheSize; i++) {
+                bufferCache.add(null);
+            }
+            size = -1;
+        }
+
+        private int getIndex(final int fragmentIndex) {
+            return fragmentIndex % bufferCache.size();
+        }
+
+        void setFragmentContext(final int fragmentIndex, final CompletableFuture<ByteBuffer> future) {
+            final int cacheIdx = getIndex(fragmentIndex);
+            final FragmentContext cachedEntry = bufferCache.get(cacheIdx);
+            if (cachedEntry == null) {
+                bufferCache.set(cacheIdx, new FragmentContext(fragmentIndex, future));
+            } else {
```

Comment:
This is simple and elegant. We may want to evaluate a simple LRU fragment "map" in the future, however, as not all read patterns are strictly sequential. It would be nice to kick out the LRU fragment, rather than the one that just happens to collide in index space. I think we can ship with this approach for now, certainly.

Reply:
Maybe make a ticket and a TODO for future work on LRU caching.

Reply:
Created issue #5061